File: Python/Remove Element.py
Remove Element


Summary

This is a pretty easy problem, but one may get confused by the term "in-place" and thought it is impossible to remove an element from the array without making a copy of the array.

Hints

Try two pointers.
Did you use the property of "the order of elements can be changed"?
What happens when the elements to remove are rare?



public int removeElement(int[] nums, int val) {
    int i = 0;
    for (int j = 0; j < nums.length; j++) {
        if (nums[j] != val) {
            nums[i] = nums[j];
            i++;
        }
    }
    return i;
}



Intuition

Now consider cases where the array contains few elements to remove. For example, 
n
u
m
s
=
[
1
,
2
,
3
,
5
,
4
]
,
v
a
l
=
4
nums=[1,2,3,5,4],val=4. The previous algorithm will do unnecessary copy operation of the first four elements. Another example is 
n
u
m
s
=
[
4
,
1
,
2
,
3
,
5
]
,
v
a
l
=
4
nums=[4,1,2,3,5],val=4. It seems unnecessary to move elements 
[
1
,
2
,
3
,
5
]
[1,2,3,5] one step left as the problem description mentions that the order of elements could be changed.


public int removeElement(int[] nums, int val) {
    int i = 0;
    int n = nums.length;
    while (i < n) {
        if (nums[i] == val) {
            nums[i] = nums[n - 1];
            // reduce array size by one
            n--;
        } else {
            i++;
        }
    }
    return n;
}

File: Java/Remove Element.java
/*
Question:
Given an array and a value, remove all instances of that value in place and return the new length.

Do not allocate extra space for another array, you must do this in place with constant memory.

The order of elements can be changed. It doesn't matter what you leave beyond the new length.

Example:
Given input array nums = [3,2,2,3], val = 3

Your function should return length = 2, with the first two elements of nums being 2.

*/


/*
This is a more efficient version which require less operations(because the order of elements can be changed) ==> different from remove zeros
*/
Solution1
public class Solution {
    public int removeElement(int[] A, int elem) {
    int l = A.length;
    for (int i=0; i<l; i++) {
        if (A[i] == elem) {
            A[i--] = A[l-- -1];
        }
    }
    return l;
}
}



/*
Two pointer solution, which is same as "REMOVE ZEROS"

*/
Solution2
public class Solution {
    public int removeElement(int[] nums, int val) {
        int p = 0;
        for(int i = 0 ; i<nums.length ; i++)
        {
            if(nums[i]!=val)
            {
                nums[p] = nums[i];
                p++;
            }
        }
        return p;
    }
}

*/


Note:

1....
2....
3....